src/mainboard/pcengines/apu1/mainboard.c

```diff
@@ -316,32 +317,42 @@ const char *smbios_mainboard_serial_number(void)
 {
 	static char serial[10];
 	struct device *dev;
-	uintptr_t bar18;
 	u32 mac_addr = 0;
+	u32 val32 = 0;
 	int i;
 
 	/* Already initialized. */
 	if (serial[0] != 0)
 		return serial;
 
 	dev = pcidev_on_root(4, 0);
+
 	if (dev)
 		dev = pcidev_path_behind(dev->link_list, PCI_DEVFN(0, 0));
 	if (!dev)
 		return serial;
 
+	/* Read PCI Configuration Space Bar 2 (0x18 offset).
+	 * Bits 31-4 contain memory address where internal NIC registers are mapped.
+	 * val32 stores that address.
+	 */
+	val32 = pci_io_read_config32(PCI_DEV(1, 0, 0), PCI_BAR_2_REG);
```

Comment:
The bar18 variable as uintptr was good. Now we introduce some mysterious val32 and introduce another mac_reg_addr varaible. This is not needed at all. Moreover, you are hardcoding the BDF of the first NIC to be 1:00.0. This will not always be true if someone connects a mPCIe module... This line `dev = pcidev_path_behind(dev->link_list, PCI_DEVFN(0, 0));` was using the PCIe bridge to detect the correct device behind it.

Reply:
I understand. The problem is when device is found with `pcidev_path_behind(dev->link_list, PCI_DEVFN(0, 0))` and `dev` is passed to `pci_read_config32` function, it always returns 0. Although `dev` is correctly detected. That's why I decided to use `pci_io_read_config32(PCI_DEV(1, 0, 0), PCI_BAR_2_REG)`. Only this function returned correct data. I will try to retrieve BDF from `dev` structure and replace the hardcoded `PCI_DEV(1, 0, 0)` with it.

Reply:
Question is what does it mean detected. Maybe it is not null. I would try to print some debug messages to verify that the structure points to the NIC

Reply:
I have done it. I have printed vendor and device id, which points to correct vendor and 'Ethernet controller' device. However, passing `dev` to mentioned function doesn't work. If there is a way to simply retrieve bus, device and function from device structure, then it might be passed instead of hardcoded values.

Reply:
Just see how struct device looks like. It should be possible with dev->bus or something